Ports the graph tracing to torch.fx

Jegp opened the pull request:
This PR ports the graph tracing to torch.fx instead of using the homebrewed graph tracer.

TBC...

Jegp committed:
Started a port of the graph tracing to torch.fx

Jegp committed:
Added tests for fx graph tracing

chanokin commented on nirtorch/graph_fx.py:
```diff
@@ -0,0 +1,207 @@
+                appear with the qualified name ``foo.bar.baz`` here.
+        """
+        # Tests that the module is in the list of custom leaves
+        if self.custom_leaf_modules and isinstance(m, self.custom_leaf_modules):
```
Do we need to check for `torch.nn.Sequential` here?

Jegp replied:
The nodes we check for here will be leaf nodes, so they won't be processed any further (which makes sense in NIR because an LI node can't be broken further down). Shouldn't Sequential nodes be processed recursively? That is, *not* be marked as leaf nodes?

chanokin replied:
Ooooh right, right. That does makes sense!

chanokin commented on nirtorch/graph_fx.py:
```diff
@@ -0,0 +1,207 @@
+    nodes = {}
+    edges = []
+    ignored_nodes = set()
+    skipped_nodes = set()
```
Does `skipped_nodes` means nodes we still need to process?

Jegp replied:
It means that we can rewire around them. That is, if we want to skip B in `A -> B -> C` we can avoid B and go straight from `A -> C`. This is useful when encountering `+` because in that case, we simply want circumvent `+` since, in NIR, two edges into the same node is exactly addition.

chanokin replied:
Right, so the only thing on this one is that the variable name is kind-a confusing but this is stricktly personal preference. To me skipped hints to the ignore these nodes path, even if you already have the `ignored_nodes` `set` there.

Jegp replied:
That's actually a good point. What would be a better name? "bypassed_nodes"? :)

chanokin replied:
I was thinking more on something in the lines of "rewiring_nodes" or "postprocessed_nodes"

Jegp replied:
What about "bypass"?

chanokin commented on nirtorch/graph_fx.py:
```diff
@@ -0,0 +1,207 @@
+                continue
+        elif node.op == "output":
+            nodes[str(node.name)] = nir.Output(np.array([1]))
+        elif node.op == "call_function":
```
Should we add a check for other non-allowed functions (multiplies or divisions) and add an exception?

Jegp replied:
Yeah! But aren't we already raising an exception if the operator is not add?

chanokin replied:
very wise, sorry about being pedantic

Jegp replied:
Glad you're checking!

chanokin commented on tests/test_graph_fx.py:
```diff
@@ -0,0 +1,190 @@
+    assert len(_filter_edges(graph, nir.Affine, nir.Output)) == 1
+
+
+def test_trace_addition():
```
how about substraction? does that work?

Jegp replied:
It doesn't! Because we don't have an equivalent in NIR. Yet? Is that something you/we would need?

chanokin replied:
Maybe? But let's just move along - it is a subject for future us to revise.

chanokin reviewed (commented):
this is great work, almost ready!

Jegp committed:
Added VSCode devcontainer

Jegp committed:
Added method to import from NIR to Torch using torch.fx

Jegp committed:
Formatted tests

Jegp commented:
I've added a method to translate NIR graphs to Torch graphs using torch.fx. There are a few questions I would be curious to hear your opinion on:

- We previously used a function to map NIR nodes to Torch nodes. I replaced that with a dictionary, where we can use the keys to check for a mapping. That is `Dict[nir.NIRNode, Callable[[nir.NIRNode], torch.nn.Module]]`. Downsides are that we have to deprecate the old syntax and the dictionary will be a little annoying to build. We'd have to have individual lambdas/functions plugged into the dictionary. Not hard, just nasty syntax. Is it worth it or should we stay with the old function approach?
- Some modules (like the ones in Norse) use stateful function calls and returns tuples of (output, state). That means, the graph should be able to handle state. By default, *all modules will work with states*. Even if there are no modules with state, we provide None values for each module, just to take care of the few modules that do. An alternative approach would be to store mutable state inside the modules. But I despise that method because we might forget to reset the state, leave it dangling in memory, etc. etc. Inputs are welcome :-)

chanokin commented on nirtorch/interpreter.py:
```diff
@@ -0,0 +1,210 @@
+    return tuple(node_outputs[a] for a, _ in input_nodes)
+
+
+def _is_stateful(module: torch.nn.Module) -> bool:
```
Given that you have an "or" at the end, does this really ensure you have a "state" argument in the function?  Just becaues `_construct_fx_graph` depends on this when you have a stateful function. 

Also, how would you enforce this in customizations to the frameworks, for example, if John Doe creates a brand new `AwesomeNeuron` type for `Norse`, can you raise an exception if the `forward` method lacks the "state" argument?

chanokin reviewed (commented):
just a tiny observation in the `_is_stateful` and `_construct_fx_graph` functions.

chanokin commented:
>     * We previously used a function to map NIR nodes to Torch nodes. I replaced that with a dictionary, where we can use the keys to check for a mapping. That is `Dict[nir.NIRNode, Callable[[nir.NIRNode], torch.nn.Module]]`. Downsides are that we have to deprecate the old syntax and the dictionary will be a little annoying to build. We'd have to have individual lambdas/functions plugged into the dictionary. Not hard, just nasty syntax. Is it worth it or should we stay with the old function approach?

I remember we talked about dictionaries making things easier to grab lists of available Nodes and to add custom nodes without needing to modify codebases for the packages (be it NIR, Norse or SNNTorch).

I'm not entirely sure what the problem with the pointer to a function is? As long as you have somewhat fixed arguments, it's no biggie in my mind.



All that being said, I think this is ready to fly! Minor adjustments could be dealt with later.

Jegp committed:
Added support for recursive interpreting. Cleaned up naming and added deprecation

Jegp committed:
Linted and formatted

pabogdan commented on nirtorch/nir_interpreter.py:
```diff
@@ -0,0 +1,285 @@
+
+        if isinstance(module, nir.Input):
+            if len(module.output_type) > 1:
+                raise ValueError("Multiple inputs are currently not supported")
```
Would this be better as a NotImplementedError?

Jegp replied:
Yes!

pabogdan reviewed (approved):
It's difficult for me to say at the moment with confidence, but this looks good to me (superficially). Mapping such as these should enable more complex extensions / transforms as well if someone needs to implement those

Jegp commented on nirtorch/nir_interpreter.py:
```diff
@@ -0,0 +1,285 @@
+    # Loop through all the nodes in the queue
+    while module_queue:
+        module_name, module = module_queue.popleft()
+        if recursion_counter[module_name] > 3:
```
3 is the third time around (note that the `collections.Counter` starts counting with 1). Technically, the worst case for the revisit is `N - 1` if you have `N` modules: they could be listed in the opposite order. The first module would wait for the second, but the second would wait for the third, and so on. We should maybe set this to `N - 1` then?

Jegp committed:
Renamed skipped -> bypass and added minor fixes from PR comments

Jegp commented:
> > ```
> > * We previously used a function to map NIR nodes to Torch nodes. I replaced that with a dictionary, where we can use the keys to check for a mapping. That is `Dict[nir.NIRNode, Callable[[nir.NIRNode], torch.nn.Module]]`. Downsides are that we have to deprecate the old syntax and the dictionary will be a little annoying to build. We'd have to have individual lambdas/functions plugged into the dictionary. Not hard, just nasty syntax. Is it worth it or should we stay with the old function approach?
> > ```
> 
> I remember we talked about dictionaries making things easier to grab lists of available Nodes and to add custom nodes without needing to modify codebases for the packages (be it NIR, Norse or SNNTorch).
> 
> I'm not entirely sure what the problem with the pointer to a function is? As long as you have somewhat fixed arguments, it's no biggie in my mind.
> 
> All that being said, I think this is ready to fly! Minor adjustments could be dealt with later.

The main problem is that we cannot see whether a module is defined without running the code. Having a set of traceable nodes is necessary to mark them as leaf nodes when processing them with torch.fx. We could let the user specify a set of modules in a separate datastructure, but that seems cumbersome...

Thank you to @pabogdan and @chanokin for the comments! If you're ok with it, I'll merge it for now. We can always improve the code later if something needs changing
